[build] add support for 2 stage rootfs build

Yakiv-Huryk opened the pull request:
This adds optimization for the SONiC image build by splitting the final build step into two stages. It allows running the first stage in parallel, improving build time.

The optimization is enabled via new rules/config flag _ENABLE_RFS_SPLIT_BUILD_ (disabled by default)

#### Why I did it
To improve a build time.

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
Added a logic to run build_debian.sh in two stages, transferring the progress via a new build artifact.

#### How to verify it
```bash
make ENABLE_RFS_SPLIT_BUILD=y SONIC_BUILD_JOBS=32 target/<IMAGE_NAME>.bin
```
#### Which release branch to backport (provide reason below if selected)

<!--
- Note we only backport fixes to a release branch, *not* features!
- Please also provide a reason for the backporting below.
- e.g.
- [x] 202006
-->

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

<!--
- Please provide tested image version
- e.g.
- [x] 20201231.100
-->

- [ ] <!-- image version 1 -->
- [ ] <!-- image version 2 -->

#### Description for the changelog
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

<!--
 Ensure to add label/tag for the feature raised. example - PR#2174 under sonic-utilities repo. where, Generic Config and Update feature has been labelled as GCU.
-->

#### Link to config_db schema for YANG module changes
<!--
Provide a link to config_db schema for the table for which YANG model
is defined
Link should point to correct section on https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/doc/Configuration.md
-->

#### A picture of a cute animal (not mandatory but encouraged)

Yakiv-Huryk commented:
HLD:
https://github.com/sonic-net/SONiC/pull/1430

Yakiv-Huryk commented:
/azpw run

mssonicbld commented:
/AzurePipelines run

Yakiv-Huryk commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

saiarcot895 commented on build_debian.sh:
```diff
@@ -616,6 +621,46 @@ if [[ ! -f './asic_config_checksum' ]]; then
+    sudo LANG=C chroot $FILESYSTEM_ROOT umount /proc
+
+    sudo rm -f $TARGET_PATH/$RFS_SQUASHFS_NAME
+    sudo mksquashfs $FILESYSTEM_ROOT $TARGET_PATH/$RFS_SQUASHFS_NAME
```
Agreed. I _think_ a simple `tar` (without any compression) would serve the purpose here.

Yakiv-Huryk replied:
Good point. I did some testing:
* default: 27 sec, 532M
* compression-level 1: 2 sec, 574M
* no compression: 1 sec, 1.4G

The "compression-level 1" seems to be the optimal tradeoff.

@saiarcot895 
Regarding `tar`, I'd stick with `mksquashfs` since it is already tested and (after compression tuning) performs very well.

I've updated the PR.

Yakiv-Huryk commented on slave.mk:
```diff
@@ -1208,6 +1210,60 @@ $(DOCKER_LOAD_TARGETS) : $(TARGET_PATH)/%.gz-load : .platform docker-start $$(TA
+$(addprefix $(TARGET_PATH)/, $(SONIC_RFS_TARGETS)) : $(TARGET_PATH)/% : \
+        .platform \
+        build_debian.sh \
+        $(addprefix $(IMAGE_DISTRO_DEBS_PATH)/,$(INITRAMFS_TOOLS) $(LINUX_KERNEL))
```
Added in https://github.com/sonic-net/sonic-buildimage/pull/15924/commits/8e1e685e3133235c49af5648595e5eacdfbc9452

Yakiv-Huryk commented on build_debian.sh:
```diff
@@ -616,6 +621,46 @@ if [[ ! -f './asic_config_checksum' ]]; then
+    ## Kill the processes
+    sudo LANG=C chroot $FILESYSTEM_ROOT fuser -km /proc || true
+    ## Wait fuser fully kill the processes
+    sleep 15
```
As you pointed out, it's derived from existing code. This is a separate issue, fixing it is out of the scope of this PR.

Yakiv-Huryk replied:
Done in https://github.com/sonic-net/sonic-buildimage/pull/15924/commits/b6f90066a59c5171cfb38651cd19ac38a3609ede

qiluo-msft commented on slave.mk:
```diff
@@ -1208,6 +1210,62 @@ $(DOCKER_LOAD_TARGETS) : $(TARGET_PATH)/%.gz-load : .platform docker-start $$(TA
+SONIC_TARGET_LIST += $(addprefix $(TARGET_PATH)/, $(SONIC_RFS_TARGETS))
+endif
+
+$(addprefix $(TARGET_PATH)/, $(SONIC_RFS_TARGETS)) : $(TARGET_PATH)/% : \
```
>SONIC_RFS_TARGETS [](http://example.com/codeflow?start=31&length=17)

Could you make sure the new target `SONIC_RFS_TARGETS` is reused if we build multiple vendors image in series?

We also have reproducible mechanism in rules/*.dep files. Is it possible to define this target's dep file, so the target is reused as cached for other later build jobs?

@xumia to review.

Yakiv-Huryk replied:
The target is reused. I've also added a cache: https://github.com/sonic-net/sonic-buildimage/pull/15924/commits/a880d53d14d593f6739e0ca33d22354157503aec
Since the targets are defined dynamically, I could not add the '.dep' file. Instead, the relevant variables are also set dynamically.

liat-grozovik commented:
who ever is reviewing this PR, we have a target to finish this review and enjoy the approved time of build in the next week.
who ever is tagged and wish to provide feedback please do it asap. 
@Yakiv-Huryk @xumia @Kalimuthu-Velappan

Kalimuthu-Velappan commented on slave.mk:
```diff
@@ -1210,21 +1210,23 @@ $(DOCKER_LOAD_TARGETS) : $(TARGET_PATH)/%.gz-load : .platform docker-start $$(TA
-
 SONIC_RFS_TARGETS= $(foreach installer, $(SONIC_INSTALLERS), $(call rfs_get_installer_dependencies,$(installer)))
+
+ifeq ($(ENABLE_RFS_SPLIT_BUILD),y)
```
Can you split the build_debian.sh into two separate files?
1. build_debian_rootfs.sh  => which contains all the standard squashfs packages
2. build_debian_sonic.sh => Includes all the sonic packages into existing squashfs and creates the dockerfs.

Yakiv-Huryk replied:
I'm not sure I understand why splitting the build_debian.sh is better. Also, with the current design, the feature is optional and configurable (via ENABLE_RFS_SPLIT_BUILD). If we split the build_debian.sh, the current way to build will no longer be available. 

Can you please help to understand why the _if..else_ is better avoided?

Kalimuthu-Velappan commented on slave.mk:
```diff
@@ -1573,7 +1575,9 @@ SONIC_CLEAN_TARGETS += $(addsuffix -clean,$(addprefix $(TARGET_PATH)/, \
 		       $(SONIC_SIMPLE_DOCKER_IMAGES) \
-		       $(SONIC_INSTALLERS)))
+		       $(SONIC_INSTALLERS) \
+		       $(SONIC_RFS_TARGETS)))
```
Can you add caching support for rootfs preparation in squashfs file system?
You can refer DPKG caching framework for deb packages.

Yakiv-Huryk replied:
Added in https://github.com/sonic-net/sonic-buildimage/pull/15924/commits/a880d53d14d593f6739e0ca33d22354157503aec

Kalimuthu-Velappan commented on slave.mk:
```diff
@@ -1210,21 +1210,23 @@ $(DOCKER_LOAD_TARGETS) : $(TARGET_PATH)/%.gz-load : .platform docker-start $$(TA
         build_debian.sh \
-        $(addprefix $(IMAGE_DISTRO_DEBS_PATH)/,$(INITRAMFS_TOOLS) $(LINUX_KERNEL))
+        $(addprefix $(IMAGE_DISTRO_DEBS_PATH)/,$(INITRAMFS_TOOLS) $(LINUX_KERNEL)) \
+        $(addsuffix -install,$(addprefix $(IMAGE_DISTRO_DEBS_PATH)/,$(DEBOOTSTRAP)))
```
Can you add tmpfs support for generating the squashfs preparation?
This will improve the performance by 5x.

Yakiv-Huryk replied:
It's a nice idea. Unfortunately, it's not in the scope of this PR.

Yakiv-Huryk commented:
> 
@k-v1 
Fixed the # 1.

Regarding the first stage dependencies (including many files from `/files`):
I've collected a list and added it to the _DEP_FILES. 
However, it's not a robust solution since there is no way to track what is used in the `build_debian.sh`, and the list should be manually maintained which is not optimal.
@xumia @Kalimuthu-Velappan  do you have any ideas/suggestions on how to handle this?

qiluo-msft commented on rules/config:
```diff
@@ -300,3 +300,6 @@ GZ_COMPRESS_PROGRAM ?= gzip
 SONIC_OS_VERSION ?= 11
+
+# ENABLE_RFS_SPLIT - enable 2-stage installer build
+ENABLE_RFS_SPLIT_BUILD ?= n
```
>ENABLE_RFS_SPLIT_BUILD [](http://example.com/codeflow?start=0&length=22)

Since it is an optimization, no need to enable it by config, just enable by default.

Yakiv-Huryk replied:
I'd suggest to consider keeping the flag and having it on by default.
In this case, if there are any issues, we can just disable it and have the build working as it is now.

@qiluo-msft @xumia @Kalimuthu-Velappan what do you think?

qiluo-msft replied:
If there are any issues, we can just revert this PR since the PR is small.

Yakiv-Huryk replied:
Ok, done

saiarcot895 commented on Makefile.cache:
```diff
@@ -357,6 +357,20 @@ define SAVE_CACHE
+	$(addprefix files/image_config/, kdump/vmcore-sysctl.conf ntp/ntp-apparmor sonic_release ntp/ntp ntp/ntp-systemd-wrapper \
+		systemd/systemd-udevd/fix-udev-rule-may-fail-if-daemon-reload-command-runs.patch) \
+	$(addprefix files/sshd/, host-ssh-keygen.sh override.conf) \
+	$(addprefix files/dhcp/, rfc3442-classless-routes sethostname sethostname6 graphserviceurl snmpcommunity vrf) \
```
Would it be better to use a wildcard/glob expression here, so that if some file gets added/removed, it's automatically added to this list?

Yakiv-Huryk replied:
Good idea, added in https://github.com/sonic-net/sonic-buildimage/pull/15924/commits/2d638f0755d103db0e979ee5b0265f692907e019

dgsudharsan commented:
@saiarcot895  @qiluo-msft Can you please review and signoff?

saiarcot895 commented on slave.mk:
```diff
@@ -1210,6 +1238,61 @@ $(DOCKER_LOAD_TARGETS) : $(TARGET_PATH)/%.gz-load : .platform docker-start $$(TA
+		export image_type="$($(installer)_IMAGE_TYPE)"
+		export sonicadmin_user="$(USERNAME)"
+		export sonic_asic_platform="$(patsubst %-$(CONFIGURED_ARCH),%,$(CONFIGURED_PLATFORM))"
+		export RFS_SPLIT_FIRST_STAGE=y
```
Nitpick, but can you set `RFS_SPLIT_LAST_STAGE` to `n` here, so that in case in the future, some code gets added that references this variable before `RFS_SPLIT_FIRST_STAGE`, it will still be defined?

Yakiv-Huryk replied:
Sure, added in https://github.com/sonic-net/sonic-buildimage/pull/15924/commits/b6f90066a59c5171cfb38651cd19ac38a3609ede

Yakiv-Huryk commented on slave.mk:
```diff
@@ -1210,6 +1238,61 @@ $(DOCKER_LOAD_TARGETS) : $(TARGET_PATH)/%.gz-load : .platform docker-start $$(TA
+		export sonic_asic_platform="$(patsubst %-$(CONFIGURED_ARCH),%,$(CONFIGURED_PLATFORM))"
+		export RFS_SPLIT_FIRST_STAGE=y
+
+		j2 -f env files/initramfs-tools/union-mount.j2 onie-image.conf > files/initramfs-tools/union-mount
+		j2 -f env files/initramfs-tools/arista-convertfs.j2 onie-image.conf > files/initramfs-tools/arista-convertfs
```
Make sense. Anyway, I've fixed it to use the `git ls-files`

saiarcot895 commented on Makefile.cache:
```diff
@@ -357,6 +357,21 @@ define SAVE_CACHE
+	$(shell git ls-files files/apparmor) \
+	$(shell git ls-files files/apt) \
+	$(wildcard files/sshd/*) \
+	$(wildcard files/dhcp/*) \
```
Can these be changed to `git ls-files` as well?

Yakiv-Huryk replied:
Sure, done.

Yakiv-Huryk commented:
/azpw run Azure.sonic-buildimage

Yakiv-Huryk commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

liat-grozovik commented:
@saiarcot895 could you please take further look now that the comments were addressed?

liat-grozovik commented:
@Yakiv-Huryk many thanks for this great achievements. 
Thanks for all the rest who provided feedback and made it even better

skg-net commented:
@Yakiv-Huryk Can you please update the Quality Metric (Alpha/Beta/GA) for the feature either in this PR comments or in HLD itself based on https://github.com/sonic-net/SONiC/blob/master/doc/SONiC%20feature%20quality%20definition.md
Thanks

Yakiv-Huryk commented:
> @Yakiv-Huryk Can you please update the Quality Metric (Alpha/Beta/GA) for the feature either in this PR comments or in HLD itself based on https://github.com/sonic-net/SONiC/blob/master/doc/SONiC%20feature%20quality%20definition.md Thanks

It’s GA. Since it’s a build improvement, it doesn't really fit the Feature Quality Definition. There is no flag to enable/disable it and no sonic-mgmt tests.
